### PR TITLE
Tensorflow graph mode

### PR DESCRIPTION
**Pull request by mikhailnikolaev**
GMM, Hand, BA, and LSTM objective test classes, using Tensorflow in graph mode, are implemented. Common code, that is used both by eager TF test classes and graph TF test classes, is moved to the directory `\src\python\modules\TensorflowCommon`.

# Graph examples
**Note**: plots for tools other than `Tensorflow`, `Manual`, and `TensorflowGraph` could be wrong, becuase they use old computed values.

## BA
BA  Jacobian  - Release Graph
BA  Jacobian ÷ objective  - Release Graph
BA  Objective  - Release Graph
BA  Objective ÷ manual  - Release Graph

## GMM
GMM (10k)  Jacobian  - Release Graph
GMM (1k)  Jacobian  - Release Graph

GMM (10k)  Jacobian ÷ objective  - Release Graph
GMM (1k)  Jacobian ÷ objective  - Release Graph

GMM (10k)  Objective  - Release Graph
GMM (1k)  Objective  - Release Graph

GMM (10k)  Objective ÷ manual  - Release Graph
GMM (1k)  Objective ÷ manual  - Release Graph

## D-LSTM
D-LSTM  Jacobian  - Release Graph

D-LSTM  Jacobian ÷ objective  - Release Graph

D-LSTM  Objective  - Release Graph

D-LSTM  Objective ÷ manual  - Release Graph

## Hand
HAND (Simple, Big)  Jacobian  - Release Graph
HAND (Simple, Small)  Jacobian  - Release Graph
HAND (Complicated, Big)  Jacobian  - Release Graph
HAND (Complicated, Small)  Jacobian  - Release Graph

HAND (Simple, Big)  Jacobian ÷ objective  - Release Graph
HAND (Simple, Small)  Jacobian ÷ objective  - Release Graph
HAND (Complicated, Big)  Jacobian ÷ objective  - Release Graph
HAND (Complicated, Small)  Jacobian ÷ objective  - Release Graph

HAND (Simple, Big)  Objective  - Release Graph
HAND (Simple, Small)  Objective  - Release Graph
HAND (Complicated, Big)  Objective  - Release Graph
HAND (Complicated, Small)  Objective  - Release Graph

HAND (Complicated, Small)  Objective ÷ manual  - Release Graph
HAND (Simple, Big)  Objective ÷ manual  - Release Graph
HAND (Simple, Small)  Objective ÷ manual  - Release Graph
HAND (Complicated, Big)  Objective ÷ manual  - Release Graph

### CODE REVIEW

File: ADBench/run-all.ps1 [hunk 1]
```diff
@@ -631,6 +631,7 @@ $tool_descriptors = @(
     [Tool]::new("Tapenade", "cpp", [ObjectiveType] "BA, LSTM, GMM, Hand", $true, $default_tolerance)
     [Tool]::new("PyTorch", "python", [ObjectiveType] "BA, LSTM, GMM, Hand", $true, 1e-7)
     [Tool]::new("Tensorflow", "python", [ObjectiveType] "BA, LSTM, GMM, Hand", $true, $default_tolerance)
+    [Tool]::new("TensorflowGraph", "python", [ObjectiveType] "BA, LSTM, GMM, Hand", $true, $default_tolerance)
     [Tool]::new("Autograd", "py", [ObjectiveType] "GMM, BA", $false, 0.0, $true, $false)
```
Comment (awf) [inline ADBench/run-all.ps1]: Is "py" different from "python"?   Not this PR I understand, but just checking.

File: src/python/modules/TensorflowGraph/TensorflowGraphBA.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        ''' Calculates objective function jacobian many times.'''
+
+        for _ in range(times):
+            # calculate reprojection and weight error derivatives part by
+            # part. Note, that weight error should be added to the sparse
+            # matrix only after the last reprojection error jacobian part
+            # adding, otherwise the result will be wrong
+            dws = []
+            for i in range(self.p):
+                dr, dw = self.session.run(
```
Comment (awf) [inline src/python/modules/TensorflowGraph/TensorflowGraphBA.py]: I guess there is a sparse matrix gather which might allow us to do this for all points at once?

That's fine as future work (i.e. create an issue) if so.

File: ADBench/run-all.ps1 (hunk 1 above)
Comment (mikhailnikolaev) [reply ADBench/run-all.ps1]: Yes, "py" is using for tools that are not from `src` directory, they are not run by the python runner, actually

File: src/python/modules/TensorflowGraph/TensorflowGraphBA.py (hunk 2 above)
Comment (mikhailnikolaev) [reply src/python/modules/TensorflowGraph/TensorflowGraphBA.py]: The sparse matrix that is used now in all python midules was written by me, using C++ realization as a pattern. So, the current realization (of both python sparse matrix and C++ one) doesn't allow adding a batch of blocks. If you meant that calculating all jacobian blocks at once could be better, then I can say that it couldn't. Firstly, creating such a huge computational graph takes a lot of operating memory. Secondly, my experiments show, that the current realization works faster

Comment (awf) [reply src/python/modules/TensorflowGraph/TensorflowGraphBA.py]: Thanks, great to know.

**Pull request merged**